lib/storage/data/external/GCP/GcpService.js

```diff
@@ -1,141 +1,889 @@
+ *
+ * Configuration options:
+ *   - s3Params: AWS SDK v3 S3Client configuration (endpoint, credentials, etc.)
+ *   - bucketName: Main GCP bucket name
+ *   - mpuBucket: Bucket for multipart uploads (optional, defaults to bucketName)
+ *   - dataStoreName: Logical name for the backend
+ *   - type: Should be 'gcp'
+ *
+ * GCP-specific behaviors:
+ *   - Object tags are stored as metadata with the prefix 'scal-s3b-tag-'.
+ *   - MPU (multipart upload) is emulated using GCP's compose/copy APIs.
+ *   - All S3-compatible methods are available, with overrides for GCP quirks.
+ */
+class GcpClient extends S3Client {
+    /**
+     * @param {object} config - Configuration object (see above)
+     */
+    constructor(config) {
+        super(config.s3Params);
+        this._config = config;
+        this._maxConcurrent = 10;
 
-Object.assign(GCP.prototype, GcpServiceSetup, {
-    _maxConcurrent: 5,
+        this.middlewareStack.remove('httpSigningMiddleware');
 
-    // Implemented APIs
-    // Bucket API
-    getBucket(params, callback) {
-        return this.listObjects(params, callback);
-    },
+        // Normalize headers before signing to avoid non-string values
+        this.middlewareStack.add(
+            createNormalizeHeadersMiddleware(),
+            {
+                name: 'normalizeHeadersMiddleware',
+                step: 'build',
+                priority: 'high',
+            }
+        );
 
-    // Object APIs
-    upload(params, callback) {
-        try {
-            const uploader = new GcpManagedUpload(this, params);
-            return uploader.send(callback);
-        } catch (err) {
-            return callback(err);
-        }
-    },
+        // Add DNS compatibility middleware AFTER bucket endpoint resolution
+        // This runs after the SDK decides virtual-hosted vs path-style
+        // and forces path-style for DNS-incompatible buckets
+        this.middlewareStack.add(
+            gcpDnsCompatibilityMiddleware(config),
+            {
+                name: 'gcpDnsCompatibility',
+                step: 'serialize',
+                priority: 'high'
+            }
+        );
+
+        // Add GCP signing middleware
+        this.middlewareStack.add(
+            gcpSigningMiddleware(config),
+            {
+                name: 'gcpSigningMiddleware',
+                step: 'finalizeRequest',
+                priority: 'high'
+            }
+        );
+    }
 
-    putObjectCopy(params, callback) {
-        return this.copyObject(params, callback);
-    },
+    /**
+     * Upload an object to GCP (with optional tagging as metadata).
+     * @param {object} params - S3 putObject params
+     * @param {function} callback
+     */
+    putObject(params, callback) {
+        const command = new PutObjectCommand(params);
+        
+        // Capture GCP-specific headers
+        let objectGeneration;
+        command.middlewareStack.add(
```

Comment:
it seems we add this middleware to every command : can't we simply add it to the `client` middleware stack instead?

c.f. https://aws.amazon.com/blogs/developer/middleware-stack-modular-aws-sdk-js/ exemple:

```
const { S3 } = require("@aws-sdk/client-s3");
const client = new S3({ region: "us-west-2" });
// Middleware added to client, applies to all commands.
client.middlewareStack.add(
  (next, context) => async (args) => {
    args.request.headers["x-amz-meta-foo"] = "bar";
    const result = await next(args);
    // result.response contains data returned from next middleware.
    return result;
  },
  {
    step: "build",
    name: "addFooMetadataMiddleware",
    tags: ["METADATA", "FOO"],
  }
);

await client.putObject(params);
```

Reply:
yes, this logic must be made reusable, no point repeating it.